docs/storage.md

```diff
@@ -0,0 +1,57 @@
+
+More ledgers can added by plugins. Each correct node should have exactly the same transactions for each ledger id.
+
+A ledger is associated with a [compact merkle tree](https://github.com/google/certificate-transparency/blob/master/python/ct/crypto/merkle.py). 
+Each new transaction is added to the ledger (log) and is also hashed (sha256) and this hash becomes a new leaf of the merkle tree which also 
+results in a new merkle root. Thus for each transaction a merkle proof of presence, called `inclusion proof` or `audit_path` can be created by 
+using the root hash a few (`O(lgn)`, `n` being the total number of leaves/transactions in the tree/ledger) intermediate hashes. Hash of all the 
+leaves and intermediate nodes of the tree are stored in a `HashStore`, enabling the creating of inclusion proofs and subset proofs. A subset proof 
+proves that a particular merkle tree is a subset of another (usually larger) merkle tree; more on this in the Catchup doc. The `HashStore` has 2 separate storages for storing leaves 
+and intermediate nodes, each leaf or node of the tree is 32 bytes. Each of these storages can be a binary file or a key value store. 
+The leaf or node hashes are queried by their number. When a client write request completes or it requests a transaction with a particular sequence number from a ledger, 
+the transaction is returned with its inclusion proof. 
+
+Relevant code:
+Ledger: `plenum/common/ledger.py` and `ledger/ledger.py`
+Compact Merkle Tree: `ledger/compact_merkle_tree.py`
+HashStore: `ledger/hash_stores/hash_store.py`
+
+
+#### 2. State
+The state can be considered a collection of key value pairs but this collection has some properties of merkle tree like a root hash and 
+inclusion proof of the keys, eg. If there are 3 key value pairs `k1-v1`, `k2->v2` and `k3->v3`, then state will have a root hash say `R` 
+and a proof can be generated that state with root hash `R` has a key `k1` with value `v1`, similarly for other keys. When a new key is added 
+or value of an old key is changed the root hash changes from `R` to `R'`. Note that the order of insertion of keys does not matter, any order of the 
+keys will result in the same root hash and same inclusion proof. The underlying data structure of state is the [Merkle Patricia Trie](https://blog.ethereum.org/2015/11/15/merkling-in-ethereum/) used by Ethereum.
+The state is built from a ledger, hence each ledger will usually have a corresponding state. State can be reconstructed from the Ledger.
+
+Relevant code:
+State: `state/pruning_state.py`
+Merkle Patricia Trie: `state/trie/pruning_trie.py`
+
+
+#### 3. Storage abstractions
```

Comment:
I think this Section should be independent one (not number 3).

I would create the following structure for the doc:

# Storages
### 1. Ledger
### 2. State
### 3. SeqNo DB
---- other storages

# Storage Implementation
Talk about storage abstractions here

Reply:
Done